### PR TITLE
chore: add cjs 

### PR DESCRIPTION
**Pull request by vasco-santos**
## Goal

This PR setups up a JS build using [ipjs](https://github.com/mikeal/ipjs) similarly to what happens in [multiformats](https://github.com/multiformats/js-multiformats) + [ipld/js-car](https://github.com/ipld/js-car) repos.

I am not entirely sure that `ipjs` provides the best solution for the building issue, but we can just use what we already know in the ecosystem now and rethink later.

## Project/Build Structure

For simplicity:
- project directory structure was changed to have `lib` and `test`
- manual created typescript declaration files were removed, as we can simply generate them

The resulting build structure is:

image

and we use the package.json exports property to hint how we can import things without the entire path. There is some magic provided by `ipbjs`, where the exports are updated in the `package.json` to also include `browser` and `require`. Example:

```js
".": {
      "browser": "./esm/lib/index.js",
      "require": "./cjs/lib/index.js",
      "import": "./esm/lib/index.js"
    },
```

## Types

By relying on the actual code to generate the types, some previously incorrect jsdoc were also fixed. The generics are now using simple `any`, but this should really be fixed

## Testing

The previous tests were kept, but now run in the context of cjs and esm (for Node). There is room for improvement here, by both:
- adding tests for the typescript usage (multiformats + car also do it)
  - I did not add them now because this causes issues with the binary, need to find a proper solution
  - given that the `ts-use` test is used widely https://github.com/multiformats/js-multiformats/tree/master/test/ts-use I am considering just to create a small tool instead of having to replicate this everywhere
  - (EDIT: Tested in a local typescript project and everything looks good from ts standpoint)
- adding browser tests with playwright

I would like to add them in follow up PRs

## Releasing

This now needs to be released within the dist folder, or use the release script

### CODE REVIEW

**Comment by vasco-santos** [comment]
Please notice this should be handled as a breaking change

File: lib/treewalk/splitter.js [hunk 1]
```diff
@@ -71,12 +77,17 @@ export class TreewalkCarSplitter {
...
       for await (const val of this._cars(cid, parents, channel)) {
         channel = val.channel
         yield val
       }
     }
 
+    if (!channel) {
+      throw new Error('no blocks in the car')
+    }
+
```
Comment (alanshaw) [inline lib/treewalk/splitter.js:90]: This one is unnecessary - there is always a channel (L66)

```suggestion
```

File: lib/treewalk/splitter.js [hunk 2]
```diff
@@ -33,6 +34,11 @@ export class TreewalkCarSplitter {
       channel = val.channel
       if (val.out) yield val.out
     }
+
+    if (!channel) {
+      throw new Error('no blocks in the car')
```
Comment (alanshaw) [inline lib/treewalk/splitter.js:39]: This is impossible, `this._cars` will yield a channel or throw. Lets change the error message here so if we ever change the code and it does happen then it's a bit less confusing.

```suggestion
      throw new Error('missing CAR writer channel')
```

**Review by alanshaw — CHANGES_REQUESTED** [review]
Super cool 👍 - I've just left some minor feedback but this should be good to go when addressed!

File: lib/treewalk/splitter.js (hunk 2 above)
Comment (vasco-santos) [reply lib/treewalk/splitter.js:39]: sounds good the message error. I know that a channel will always exist, but typescript catches this as an error... Other option would be to

File: lib/treewalk/splitter.js [hunk 3]
```diff
@@ -71,12 +77,17 @@ export class TreewalkCarSplitter {
...
     for (const [, cid] of block.links()) {
+      // @ts-ignore TS does not get type of async generator
       for await (const val of this._cars(cid, parents, channel)) {
         channel = val.channel
         yield val
       }
     }
 
+    if (!channel) {
+      throw new Error('no blocks in the car')
```
Comment (vasco-santos) [inline lib/treewalk/splitter.js:88]:
```suggestion
      throw new Error('missing CAR writer channel')
```

File: lib/treewalk/splitter.js (hunk 1 above)
Comment (vasco-santos) [reply lib/treewalk/splitter.js]: ts complaint yah, I added the suggested error message

**Comment by vasco-santos** [comment]
@alanshaw I removed the prepare script and prepublish as they were problematic when trying to release. 

I was using prepare for local testing in another repo

The prepublish multiformats https://github.com/multiformats/js-multiformats/blob/master/package.json does not have it and `ipjs` already runs the build before per https://github.com/mikeal/ipjs/blob/master/cli.js#L58

**Comment by vasco-santos** [comment]
I released `@vascosantos/carbites` and trying it out in nft.storage, needed to push one more types fix. As we use folders for simple/rooted/treewalk, typescript compiler is not intelligent to check if there is a `index.d.ts` in the path, which means that:

```js
import { TreewalkCarSplitter } from '@vascosantos/carbites/treewalk'
```

would not find the types. This would only work if treewalk was a file as it would try to get `treewalk.d.ts`. So, I added the explicit type path for them

**Review by alanshaw — APPROVED** [review]

**Pull request merged**